### PR TITLE
feat: send project event

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: sdk/event_project.go [hunk 1]
```diff
@@ -1,5 +1,23 @@
 package sdk
 
+// EventAddProject represents the event when adding a project
+type EventAddProject struct {
+	ProjectKey  string            `json:"projectKey"`
```
Comment (yesnault) [inline sdk/event_project.go]: project_key

File: sdk/event_project.go [hunk 2]
```diff
@@ -1,5 +1,23 @@
...
+	Variables   []Variable        `json:"variables"`
+	Permissions []GroupPermission `json:"groups_permission"`
+	Keys        []ProjectKey      `json:"keys"`
+	Metadata    Metadata          `json:"metadata"`
+}
+
+// EventUpdateProject represents the event when updating a project
+type EventUpdateProject struct {
+	ProjectKey  string   `json:"projectKey"`
```
Comment (yesnault) [inline sdk/event_project.go]: project_key

File: sdk/event_project.go [hunk 3]
```diff
@@ -57,14 +75,14 @@ type EventDeleteProjectKey struct {
 
 // EventAddProjectVCSServer represents the event when adding a project vcs server
 type EventAddProjectVCSServer struct {
-	ProjectKey string           `json:"projectKey"`
-	VCSServer  ProjectVCSServer `json:"vcs_server"`
+	ProjectKey    string `json:"projectKey"`
```
Comment (yesnault) [inline sdk/event_project.go]: project_key

File: sdk/event_project.go [hunk 4]
```diff
@@ -57,14 +75,14 @@ type EventDeleteProjectKey struct {
...
+	VCSServerName string `json:"vcs_server"`
 }
 
 // EventDeleteProjectVCSServer represents the event when deleting a project vcs server
 type EventDeleteProjectVCSServer struct {
-	ProjectKey string           `json:"projectKey"`
-	VCSServer  ProjectVCSServer `json:"vcs_server"`
+	ProjectKey    string `json:"projectKey"`
```
Comment (yesnault) [inline sdk/event_project.go]: project_key

File: engine/api/project_group.go [hunk 5]
```diff
@@ -307,6 +269,8 @@ func (api *API) addGroupInProjectHandler() Handler {
 			return sdk.WrapError(err, "AddGroupInProject: Cannot commit transaction")
 		}
 
+		go event.PublishAddProjectPermission(p, groupProject, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_group.go]: no goroutine

File: engine/api/project_key.go [hunk 6]
```diff
@@ -97,21 +98,26 @@ func (api *API) deleteKeyInProjectHandler() Handler {
...
 				}
+				break
 			}
 		}
 
 		if err := tx.Commit(); err != nil {
 			return sdk.WrapError(err, "deleteKeyInProjectHandler> Cannot commit transaction")
 		}
 
+		go event.PublishDeleteProjectKey(p, deletedKey, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_key.go]: no goroutine

File: engine/api/project_key.go [hunk 7]
```diff
@@ -173,6 +179,8 @@ func (api *API) addKeyInProjectHandler() Handler {
 			return sdk.WrapError(err, "addKeyInProjectHandler> Cannot commit transaction")
 		}
 
+		go event.PublishAddProjectKey(p, newKey, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_key.go]: no goroutine

File: engine/api/project_platform.go [hunk 8]
```diff
@@ -80,6 +81,8 @@ func (api *API) putProjectPlatformHandler() Handler {
 			return sdk.WrapError(err, "putProjectPlatformHandler> Cannot commit transaction")
 		}
 
+		go event.PublishUpdateProjectPlatform(p, ppBody, ppDB, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_platform.go]: no goroutine

File: engine/api/project_platform.go [hunk 9]
```diff
@@ -117,6 +122,8 @@ func (api *API) deleteProjectPlatformHandler() Handler {
 		if err := tx.Commit(); err != nil {
 			return sdk.WrapError(err, "deleteProjectPlatformHandler> Cannot commit transaction")
 		}
+
+		go event.PublishDeleteProjectPlatform(p, deletedPlatform, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_platform.go]: no goroutine

File: engine/api/project_platform.go [hunk 10]
```diff
@@ -177,6 +184,8 @@ func (api *API) postProjectPlatformHandler() Handler {
 			return sdk.WrapError(err, "postProjectPlatformHandler> Cannot commit transaction")
 		}
 
+		go event.PublishAddProjectPlatform(p, pp, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_platform.go]: no goroutine

File: engine/api/project_variable.go [hunk 11]
```diff
@@ -163,103 +100,9 @@ func (api *API) deleteVariableFromProjectHandler() Handler {
...
-		}
-
-		if err := project.UpdateLastModified(tx, api.Cache, getUser(ctx), p, sdk.ProjectVariableLastModificationType); err != nil {
-			return sdk.WrapError(err, "updateVariablesInProjectHandler> Cannot update last modified")
-		}
-
-		if err := tx.Commit(); err != nil {
-			return sdk.WrapError(err, "updateVariablesInProjectHandler> Cannot commit transaction")
-		}
+		go event.PublishDeleteProjectVariable(p, *varToDelete, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_variable.go]: no goroutine

File: engine/api/project_variable.go [hunk 12]
```diff
@@ -305,31 +149,9 @@ func (api *API) updateVariableInProjectHandler() Handler {
...
-			return sdk.WrapError(err, "getVariableInProjectHandler: Cannot load project %s", key)
-
-		}
-
-		v, err := project.GetVariableInProject(api.mustDB(), proj.ID, varName)
-		if err != nil {
-			return sdk.WrapError(err, "getVariableInProjectHandler: Cannot get variable %s in project %s", varName, key)
-
-		}
+		go event.PublishUpdateProjectVariable(p, newVar, *previousVar, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_variable.go]: no goroutine

File: engine/api/project_variable.go [hunk 13]
```diff
@@ -390,6 +212,9 @@ func (api *API) addVariableInProjectHandler() Handler {
 			return sdk.WrapError(err, "addVariableInProjectHandler: cannot commit tx")
 		}
 
+		// Send Add variable event
+		go event.PublishAddProjectVariable(p, newVar, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_variable.go]: no goroutine

File: engine/api/repositories_manager.go [hunk 14]
```diff
@@ -163,6 +164,8 @@ func (api *API) repositoriesManagerOAuthCallbackHandler() Handler {
 			return sdk.WrapError(errT, "repositoriesManagerAuthorizeCallback> Cannot commit transaction")
 		}
 
+		go event.PublishAddVCSServer(proj, vcsServerForProject.Name, getUser(ctx))
```
Comment (fsamin) [inline engine/api/repositories_manager.go]: no goroutine

File: engine/api/repositories_manager.go [hunk 15]
```diff
@@ -236,6 +239,8 @@ func (api *API) repositoriesManagerAuthorizeCallbackHandler() Handler {
...
```
Comment (fsamin) [inline engine/api/repositories_manager.go]: no goroutine

File: engine/api/repositories_manager.go [hunk 16]
```diff
@@ -275,6 +280,8 @@ func (api *API) deleteRepositoriesManagerHandler() Handler {
 			return sdk.WrapError(err, "deleteRepositoriesManagerHandler> Cannot commit transaction")
 		}
 
+		go event.PublishDeleteVCSServer(p, vcsServer.Name, getUser(ctx))
```
Comment (fsamin) [inline engine/api/repositories_manager.go]: no goroutine

File: engine/api/project.go [hunk 17]
```diff
@@ -114,7 +115,7 @@ func (api *API) updateProjectHandler() Handler {
 		if errUp := project.Update(api.mustDB(), api.Cache, proj, getUser(ctx)); errUp != nil {
 			return sdk.WrapError(errUp, "updateProject> Cannot update project %s", key)
 		}
-
+		go event.PublishUpdateProject(proj, p, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project.go]: no goroutine

File: engine/api/project.go [hunk 18]
```diff
@@ -309,6 +310,8 @@ func (api *API) addProjectHandler() Handler {
 			return sdk.WrapError(err, "addProjectHandler> Cannot commit transaction")
 		}
 
+		go event.PublishAddProject(p, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project.go]: no goroutine

File: engine/api/project.go [hunk 19]
```diff
@@ -347,6 +350,9 @@ func (api *API) deleteProjectHandler() Handler {
 		if err := tx.Commit(); err != nil {
 			return sdk.WrapError(err, "deleteProject> Cannot commit transaction")
 		}
+
+		go event.PublishDeleteProject(p, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project.go]: no goroutine

File: engine/api/project_group.go [hunk 20]
```diff
@@ -54,6 +66,8 @@ func (api *API) deleteGroupFromProjectHandler() Handler {
 			return sdk.WrapError(err, "deleteGroupFromProjectHandler: Cannot commit transaction")
 		}
 
+		go event.PublishDeleteProjectPermission(p, gp, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_group.go]: no goroutine

File: engine/api/project_group.go [hunk 21]
```diff
@@ -125,68 +141,14 @@ func (api *API) updateGroupRoleOnProjectHandler() Handler {
...
-			return sdk.WrapError(err, "updateGroupsInProject> unable to unmarshal")
-		}
 
-		if len(groupProject) == 0 {
-			return sdk.WrapError(sdk.ErrGroupNeedWrite, "updateGroupsInProject: Cannot remove all groups.")
+		newGP := sdk.GroupPermission{
+			Permission: groupProject.Permission,
+			Group:      gpInProject.Group,
 		}
+		go event.PublishUpdateProjectPermission(p, newGP, gpInProject, getUser(ctx))
```
Comment (fsamin) [inline engine/api/project_group.go]: no goroutine

**Pull request merged**